File: nmfp/Cpp/polynomial_generate.cpp
// Program to generate a data file containing x and y values
// of the given polynomial between user-specified limits.
//
// Author:  Edward J. Brash
//

#include <iostream>
#include <fstream>
#include "math.h"
#include <TRandom.h>

using namespace std;

int main() {

  //* Initialize the polynomial

//  double pi = 3.14159265;
//  double a0 = -pi*pi*pi;
//  double a1 = 3*pi*pi;
//  double a2 = -3*pi;
//  double a3 = 1.0;

  double a0 = -6;
  double a1 = 11;
  double a2 = -6;
  double a3 = 1;

  double xlow,xhigh;
  int npoints, i;

  ofstream myfile;
  myfile.open ("poly.dat");
  ofstream dfile;
  dfile.open ("deriv.dat");

  cout << "Enter the number of data points:" << endl;
  cin >> npoints;
  cout << "Enter the lower range:" << endl;
  cin >> xlow;
  cout << "Enter the upper range:" << endl;
  cin >> xhigh;

  const int n = npoints;
  TRandom r;

  float x0[n],d[n],f[n],ef[n];

  double step = (xhigh - xlow)/npoints;
  double rand = 0;

  for (i=0; i<npoints; i++) {
	  x0[i] = xlow + i*step;
	  f[i] = a0 + a1*x0[i] + a2*x0[i]*x0[i] + a3*x0[i]*x0[i]*x0[i];
	  ef[i]=pow(fabs(f[i]),0.5);
	  rand = ef[i];
	  f[i]=f[i]+r.Gaus(0.0,rand);
	  cout << "X-value: " << x0[i] << " F(x): " << f[i] << endl; 
	  myfile << x0[i] << " " << f[i] << " " << ef[i] << endl;
  }	  

  for (i=1;i<npoints-1; i++){
	  d[i]=(f[i+1]-f[i-1])/(2.0*step);
	  cout << "X-value: " << x0[i] << " F'(x): " << d[i] << endl; 
	  dfile << x0[i] << " " << d[i] << endl;
  }

  myfile.close();
  dfile.close();
  return 0;
	  
}

File: JupyterNotebooks/Fitting/polynomial_generate.py
#! python
# Program to generate a data file containing x and y values
# of the given polynomial between user-specified limits.
# 
# Author:  Edward J. Brash
# 

from numpy import *

pi = 3.14159265

a0 = -pi*pi*pi
a1 = 3*pi*pi
a2 = -3*pi
a3 = 1.0

poly = open("poly.dat", "w")
deriv = open("deriv.dat", "w")

npoints = int(raw_input("Enter the number of data points:"))
xlow = float(raw_input("Enter the lower range:"))
xhigh = float(raw_input("Enter the upper range:"))

n = npoints-1
x = [0]*(n)
xd = [0]*(n-2)
d = [0]*(n-2)
f = [0]*(n)
step = (xhigh - xlow)/npoints

for i in range(0,n):
    x[i] = xlow + i*step
    f[i] = a0 + a1*x[i] + a2*x[i]*x[i] + a3*x[i]*x[i]*x[i]
    print "X-value: ",x[i]," F(x): ",f[i]
    poly.write("%s %s\n" %(x[i],f[i]))

for i in range(1,n-1):
    d[i-1]=(f[i+1]-f[i-1])/(2.0*step)
    xd[i-1]=x[i]
    print "X-value: ",xd[i-1],"F`(x): ",d[i-1]
    deriv.write("%s %s\n" %(xd[i-1],d[i-1]))

poly.close()
deriv.close()
